Library/OcCpuLib/FrequencyDetect.c

```diff
@@ -225,6 +231,94 @@ InternalCalculateTSCFromPMTimer (
   return TSCFrequency;
 }
 
+UINT64
+InternalCalculateTSCFromApplePlatformInfo (
+  OUT  UINT64  *FSBFrequency  OPTIONAL
+  )
+{
+  EFI_STATUS                             Status;
+  APPLE_PLATFORM_INFO_DATABASE_PROTOCOL  *PlatformInfo;
+  UINT32                                 Size;
+  VOID                                   *FsbHob;
+  UINT64                                 FsbFreq;
+
+  FsbFreq = 0;
+
+  Status = gBS->LocateProtocol (
+    &gApplePlatformInfoDatabaseProtocolGuid,
+    NULL,
+    (VOID **) &PlatformInfo
+    );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_VERBOSE, "OCCPU: Failed to locate ApplePlatformInfo protocol - %r\n", Status));
+    return 0;
+  }
+
+  Status = PlatformInfo->GetFirstDataSize (
+    PlatformInfo,
+    &gAppleFsbFrequencyPlatformInfoGuid,
+    &Size
+    );
+  if (!EFI_ERROR (Status)) {
+    if (Size > sizeof (UINT64) || Size < sizeof (UINT32)) {
```

Comment:
On failure this should fallback to GetFirstGuidHob (NULL) path. Perhaps you need to introduce two functions that would just read values from different places?

```C
EFI_STATUS
ReadApplePlatformFirstData (
  IN APPLE_PLATFORM_INFO  *PlatformInfo,
  IN OUT UINT32 *Size,
  OUT VOID *Data
  );
```

And another one for ReadApplePlatformData.